rendering/src/renderer.rs

```diff
@@ -0,0 +1,364 @@
+use std::iter;
+use std::mem;
+
+use pollster::FutureExt;
+use raw_window_handle::HasRawWindowHandle;
+use wgpu::util::DeviceExt;
+
+use crate::model::ModelIndex;
+use crate::model::ModelManager;
+use crate::model::{self, DrawModel, Vertex};
+use crate::{camera, texture, Camera, RenderingError};
+use common::{Mat4, Transform, Vec3, Vec4};
+
+#[repr(C)]
+#[derive(Debug, Copy, Clone, bytemuck::Pod, bytemuck::Zeroable)]
+pub struct RawTranslationMatrix {
+    model: [[f32; 4]; 4],
+}
+
+impl RawTranslationMatrix {
+    pub fn new(transform: Transform) -> Self {
+        let Vec3 { x, y, z } = transform.scale;
+
+        Self {
+            model: unsafe {
+                mem::transmute::<Mat4, _>(
+                    Mat4::translation_3d(transform.position)
+                        * Mat4::from(transform.rotation)
+                        * Mat4::with_diagonal(Vec4::new(x, y, z, 1.0)),
+                )
+            },
+        }
+    }
+}
+
+impl RawTranslationMatrix {
+    fn desc<'a>() -> wgpu::VertexBufferLayout<'a> {
+        wgpu::VertexBufferLayout {
+            array_stride: mem::size_of::<RawTranslationMatrix>() as wgpu::BufferAddress,
+            step_mode: wgpu::VertexStepMode::Instance,
+            attributes: &[
+                wgpu::VertexAttribute {
+                    offset: 0,
+                    shader_location: 5,
+                    format: wgpu::VertexFormat::Float32x4,
+                },
+                wgpu::VertexAttribute {
+                    offset: 4 * 4,
+                    shader_location: 6,
+                    format: wgpu::VertexFormat::Float32x4,
+                },
+                wgpu::VertexAttribute {
+                    offset: 2 * 4 * 4,
+                    shader_location: 7,
+                    format: wgpu::VertexFormat::Float32x4,
+                },
+                wgpu::VertexAttribute {
+                    offset: 3 * 4 * 4,
+                    shader_location: 8,
+                    format: wgpu::VertexFormat::Float32x4,
+                },
+            ],
+        }
+    }
+}
+
+pub struct Renderer {
+    pub camera: Camera,
```

Comment:
Feels really scuffed to tie the camera to the renderer, but I don't know how else to do it

Reply:
I definitly think that we should have some kind of `CameraComponent` that can be added to an entity (or possibly multiple?) and then the renderer would query for those to know where to render from.

Then the `CameraComponent` would probably be in `common` i guess

Reply:
But of course the renderer would have to know about `CameraComponent`s